# Regions-of-interest (ROIs)

**Author: Hannes Almgren (h.almgren@unsw.edu.au)**

**Citation: Almgren et al. (2025). DOI: https://doi.org/10.1007/s00415-025-12920-9**

## General Instructions:

Each box in this notebook is called a **'cell'**. You can **execute** the commands in a cell by clicking on the cell and hitting **ctrl + enter** on your keyboard.

There are a few **dataformats** that are used in this tutorial: <br>
Files that end in **'.mif' and 'nii.gz'** are **volumetric images** (either showing the brain or a specific region of the brain).  <br>
Files that end in **'.tck'** are **tractography files**.  <br>
Files that end in **.txt** are just **text-files**.  <br>
Files that end in **'.csv'** are **datasheets**.  <br>

# Let's first look at the whole-brain tractogram

In [1]:
import module
await module.load('mrtrix3/3.0.4')

ModuleNotFoundError: No module named 'module'

### Load the Mrtrix3 software package:

#### Tip: If you want to see more information about an mrtrix3-command, look for command on their <a href=https://mrtrix.readthedocs.io/en/latest/reference/commands_list.html>website</a>. 

### Import the NiiVue-visualizer to display the brain images:

In [1]:
from ipyniivue import NiiVue

### Inspect the whole-brain tractogram visually:
#### We selected 10000 streamlines to display
Note: you can inspect the tractogram from different angles by clicking on the tractogram and turning it.

In [2]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/10k_streamlines_for_displaying.tck"}
    ]
)
nv

NiiVue(height=300)

### In this tutorial we only use a subset of the streamlines to speed up our analyses.
#### Usually you would generate between 10M and 70M streamlines for a whole-brain tractogram (but you would want to select a smaller subsample for visualization)

#### The command <a href="https://mrtrix.readthedocs.io/en/latest/reference/commands/tckinfo.html">tckinfo</a> shows you information about the streamlines in your file, including the number of streamlines. <em>Note that tractography files in mrtrix3 end with '.tck'</em>. 
The output of the following line should say 200000, which is the number of streamlines in our whole-brain tractogram.

In [3]:
!tckinfo ../data/diffusion/200k_streamlines_for_tutorial.tck | grep ' count:'

    count:                200000


# PART 1: Corticospinal tract (CST)

# In the next step we will segment the corticospinal tract from the whole-brain tractogram.

### <p style="text-align: justify;"> To identify the whole corticospinal tract, we have <span style="color:green;">__two inclusion Regions-Of-Interest (ROIs)__ (the cortex and the brainstem)</span> and <span style="color:red;">__two exclusion ROIs__ (the cerebellum and corpus callosum)</span>. We will only include streamlines from the whole-brain tractography that <span style="color:green;">___pass throught the brainstem ROI and end within 4mm from the cortical ROI__ (because of partial volume effects). We will <span style="color:red;">exclude streamlines that go <span style="color:red;">__through the corpus callosum and that enter (or pass through) the cerebellum__ </span> </p>

Each ROI was defined according to specific criteria: The pons-ROI was manually drawn on a T1w image right above the actual pons (transversal slice). The cerebellum was based on freesurfer output (combining white and grey matter masks from both hemispheres). The corpus callosum was drawn manually on a few mid-sagittal slices located using a T1w-image. The cortical ROI was a combination of somatosensory, motor, and premotor areas registered to our subject-specific brain image. For more information on the latter, see further down this notebook.

You can visualize these ROIs with by running the cell below (green = inclusion ROIs, red = exclusion ROIs)

In [4]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}, 
        {"path": "../data/diffusion/pons_corrected.mif"},
        {"path": "../data/diffusion/cortex_together.mif"},
        {"path": "../data/diffusion/cerebellum_maskedWMGM.mif"},
        {"path": "../data/diffusion/corpus_callosum.mif"}
    ]
)
nv.volumes[1].colormap='green'
nv.volumes[2].colormap='green'
nv.volumes[3].colormap='red'
nv.volumes[4].colormap='red'
nv

NiiVue(height=300)

#### For the next step, we will use <a href="https://mrtrix.readthedocs.io/en/latest/reference/commands/tckedit.html">tckedit</a> to select the streamlines from the whole-brain tractogram based on our ROIs. We will select streamlines that enter the brainstem, but do not cross the corpus callosum nor enter the cerebellum. <em>Note: we will select the streamlines that end in our cortical ROIs later with a different command</em>.
Note that tckedit takes a while, so let's not run it today (we have stored the output of this command in your environment). If you want to run it, please remove the hash (#) in the cell below before you run it

In [5]:
#!tckedit ../data/diffusion/200k_streamlines_for_tutorial.tck ../data/diffusion/selected_streamlines.tck -include ../data/diffusion/pons_corrected.mif -exclude ../data/diffusion/cerebellum_maskedWMGM.mif -exclude ../data/diffusion/corpus_callosum.mif -force

tckedit: [WARNING] existing output files will be overwritten
tckedit: [done]   200000 read,     5291 written


After running the above command we should have 5291 streamlines left. Once more we can use <a href="https://mrtrix.readthedocs.io/en/latest/reference/commands/tckinfo.html">tckinfo</a> to check that:

In [6]:
!tckinfo ../data/diffusion/selected_streamlines.tck | grep ' count:'

    count:                5291


#### Let's visualize those streamlines using niivue:
Once more you can inspect the tractogram from different angles by clicking on it and turning it.

In [7]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/selected_streamlines.tck"}
    ]
)
nv

NiiVue(height=300)

### Select the streamlines that end within 4mm of our cortical ROI

#### In a next step we will select the streamlines that end close to our cortical ROI (which is a combination of the motor, somatosensory, and premotor cortex). We will allow streamlines to end within 4mm from our cortical ROIs because (1) the edges of the cortex are difficult to delineate precisely because of partial volume effects, and (2) our cortical ROIs are imperfect because of registration to our subject-specific brain and won't always perfectly follow the cortex. Therefore will use the command <a href="https://mrtrix.readthedocs.io/en/latest/reference/commands/tck2connectome.html">tck2connectome</a>, which by default allows streamliunes to end within 4mm of ROIs.  <em>Note that you can change this default behavior by specifying the radius in mm after the -assignment_radial_search tag.</em> The tck2connectome-command will identify which streamlines end close to the cortex, but won't save those streamlines in separate files. The <a href="https://mrtrix.readthedocs.io/en/latest/reference/commands/connectome2tck.html">connectome2tck</a>-command does exactly that. 
**Once again, this command takes a little while to run, so we ran it for you. If you would like to run it yourself, please remove the hash (#) before the lines below.**

In [8]:
#!tck2connectome ../data/diffusion/selected_streamlines.tck ../data/diffusion/fulltract_bilateral_CST.mif ../data/diffusion/connectivity_values.csv -out_assignments ../data/diffusion/assignments_fulltract_bilateral_CST.txt -symmetric -zero_diagonal -force
#!connectome2tck ../data/diffusion/selected_streamlines.tck ../data/diffusion/assignments_fulltract_bilateral_CST.txt ../data/diffusion/fulltract_CST -nodes 2 -force

tck2connectome: [WARNING] existing output files will be overwritten
tck2connectome: [100%] Constructing connectome[0K
connectome2tck: [WARNING] existing output files will be overwritten
connectome2tck: [done] reading streamline assignments file[0K
connectome2tck: [100%] Extracting tracks from connectome[0K


*Note: The image called fulltract_bilateral_CST.mif has three values (1 to 3). We can ignore area 1 for now (it represents the cortical and subcortical areas that we did not want to include in our analyses). Area 2 represents the cortical ROI, and area 3 represents the brainstem. connectome2tck will export files that end in our cortical ROI and in the brainstem, and add the suffix '2-3' to the exported filename (for somatosensory, motor, and premotor bundles, resp.).*

**Let's count and visualize the resulting streamlines:
There should be 2152 streamlines left in our tractogram.**

In [9]:
!tckinfo ../data/diffusion/fulltract_CST2-3.tck | grep ' count:'

    count:                2152


In [10]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/fulltract_CST2-3.tck"}
    ]
)
nv

NiiVue(height=300)

### The tractogram above already looks quite good, but there are still some spurious streamlines. Let's clean it up a bit for visualization purposes by removing the connections that anterior and posterior to the CST are likely to be spurious. We drew these ROIs manually with mrview (part of the mrtrix3 software package).

In [11]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}, 
        {"path": "../data/diffusion/ROI_cleanup_lateral.mif"},
        {"path": "../data/diffusion/ROI_cleanup_anterior.mif"},
        {"path": "../data/diffusion/ROI_cleanup_posterior.mif"},
        {"path": "../data/diffusion/ROI_cleanup_posterior2.mif"}
    ]
)
nv.volumes[1].colormap='red'
nv.volumes[2].colormap='red'
nv.volumes[3].colormap='red'
nv.volumes[4].colormap='red'
nv

NiiVue(height=300)

We can use tckedit again to select streamlines that don't cross the ROIs that we drew:

In [12]:
#!tckedit ../data/diffusion/fulltract_CST2-3.tck ../data/diffusion/fulltract_CST2-3_cleaned.tck -exclude ../data/diffusion/ROI_cleanup_lateral.mif -exclude ../data/diffusion/ROI_cleanup_anterior.mif -exclude ../data/diffusion/ROI_cleanup_posterior.mif -exclude ../data/diffusion/ROI_cleanup_posterior2.mif -force

tckedit: [WARNING] existing output files will be overwritten
tckedit: [done]     2152 read,     2130 written


There should be 2130 streamlines left in our tractogram.

In [13]:
!tckinfo ../data/diffusion/fulltract_CST2-3_cleaned.tck | grep ' count:'

    count:                2130


In [14]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/fulltract_CST2-3_cleaned.tck"}
    ]
)
nv

NiiVue(height=300)

This looks already a lot cleaner

# Now let's try to divide this tract into three parts: a motor, somatosensory, and premotor part

Note that here we already have coregistered <a href=https://pubmed.ncbi.nlm.nih.gov/34146709/>Pijnenburg (2021)'s</a> cytoarchitectonic atlas to indivual brains using freesurfer's combined volumetric and surface-based registration algorihm (<a href=https://surfer.nmr.mgh.harvard.edu/fswiki/mri_cvs_register>mri_cvs_register)</a>. 
This atlas includes ROIs for the somatosensory (BA1-3), motor (BA4), and premotor regions (BA6).

Let's visualize those ROIs (green = somatosensory, blue = motor, white = premotor):

In [15]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}, 
        {"path": "../data/diffusion/somatosensory.mif"},
        {"path": "../data/diffusion/motor.mif"},
        {"path": "../data/diffusion/premotor.mif"}
    ]
)
nv.volumes[1].colormap='green'
nv.volumes[2].colormap='blue'
nv.volumes[3].colormap='white'
nv

NiiVue(height=300)

To extract the streamlines that end within 4mm of those regions, we again use the tck2connectome and connectome2tck commands, but now we use separate ROIs for somatosensory, motor, and premotor regions. 
Once more these commands take a while to run so we already saved the output file on the server. If you would like to run the commands, please remove the has (#) from each line

In [17]:
#!tck2connectome ../data/diffusion/fulltract_CST2-3_cleaned.tck ../data/diffusion/three_regions_together_bilateral_CST.mif ../data/diffusion/connectivity_values_subdivision.csv -out_assignments ../data/diffusion/assignments_bilateral_CST_subdivision.txt -symmetric -zero_diagonal -force
#!connectome2tck ../data/diffusion/fulltract_CST2-3_cleaned.tck ../data/diffusion/assignments_bilateral_CST_subdivision.txt ../data/diffusion/CST_subdivision -nodes 2,3,4 -force

tck2connectome: [WARNING] existing output files will be overwritten
tck2connectome: [100%] Constructing connectome[0K
connectome2tck: [WARNING] existing output files will be overwritten
connectome2tck: [done] reading streamline assignments file[0K
connectome2tck: [100%] Extracting tracks from connectome[0K


*Note: The image called three_regions_together_bilateral_CST.mif has four values (1 to 4). Area 1-3 represents the cortical ROIs (somatosensory, motor, and premotor, respectively), and area 4 represents the brainstem. connectome2tck will export files that end in our cortical ROIs and in the brainstem, and add the suffix '1-4', '2-4', '3-4' to the exported filename.*

## Let's now visualize our subdivision of the cortispinal tract

**First let's visualize the somatosensory part:**

In [19]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CST_subdivision1-4.tck"}
    ]
)
nv

NiiVue(height=300)

**Second let's visualize the motor part:**

In [20]:
#Display the motor part of the CST
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CST_subdivision2-4.tck"}
    ]
)
nv

NiiVue(height=300)

**Third let's visualize the premotor part:**

In [21]:
#Display the premotor part of the CST
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CST_subdivision3-4.tck"}
    ]
)
nv

NiiVue(height=300)

### As you hopefully can see, the somatosensory subdivision lies posterior to the motor part, which lies posterior to the premotor subdivision.

# Part 2: CORPUS CALLOSUM (CC)

### <p style="text-align: justify;"> We use the same commands with similar ROIs to identify the corpus callosum. We use <span style="color:green;">__two inclusion ROIs__ (the cortex and the corpus callosum)</span>, which means that we will only include streamlines from the whole-brain tractography that <span style="color:green;">__pass through the corpus callosum ROI and end within 4mm from the cortical ROI__</span> (because of partial volume effects).

**Let's visualize those ROIs**

In [22]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}, 
        {"path": "../data/diffusion/cortex_together.mif"},
        {"path": "../data/diffusion/corpus_callosum.mif"}
    ]
)
nv.volumes[1].colormap='green'
nv.volumes[2].colormap='green'
nv

NiiVue(height=300)

**We use tckedit to select the streamlines that pass through the corpus callosum**

In [23]:
#!tckedit ../data/diffusion/200k_streamlines_for_tutorial.tck ../data/diffusion/selected_streamlines_CC.tck -include ../data/diffusion/corpus_callosum.mif -force

tckedit: [WARNING] existing output files will be overwritten
tckedit: [done]   200000 read,    33875 written


**We count the number of streamlines in the resulting file (should be 33875 streamlines)**

In [24]:
!tckinfo ../data/diffusion/selected_streamlines_CC.tck | grep ' count:'

    count:                33875


**We visualize those streamlines**

In [25]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/selected_streamlines_CC.tck"}
    ]
)
nv

NiiVue(height=300)

**We select the streamlines that end within 4mm of our cortical ROI**

In [26]:
#!tck2connectome ../data/diffusion/selected_streamlines_CC.tck ../data/diffusion/fulltract_CC.mif ../data/diffusion/connectivity_values_CC.csv -out_assignments ../data/diffusion/assignments_fulltract_bilateral_CC.txt -symmetric -zero_diagonal -force
#!connectome2tck ../data/diffusion/selected_streamlines_CC.tck ../data/diffusion/assignments_fulltract_bilateral_CC.txt ../data/diffusion/fulltract_CC -nodes 2 -force

tck2connectome: [WARNING] existing output files will be overwritten
tck2connectome: [100%] Constructing connectome[0K
connectome2tck: [WARNING] existing output files will be overwritten
connectome2tck: [done] reading streamline assignments file[0K
connectome2tck: [100%] Extracting tracks from connectome[0K


*Note: The image called fulltract_CC.mif has three values (1 to 3). We can ignore area 1 for now (it represents the cortical and subcortical areas that we did not want to include in our analyses). Area 2 represents the left cortical ROI, and area 3 represents the right cortical ROI. connectome2tck will export files that end in the left cortical ROI and in the right cortical ROI, and add the suffix '2-3' to the exported filename.*

**There should be 6658 streamlines in the streamline file:**

In [27]:
!tckinfo ../data/diffusion/fulltract_CC2-3.tck | grep ' count:'

    count:                6658


In [28]:
nv=NiiVue()
nv.load_meshes(
    [
        {"path": "../data/diffusion/fulltract_CC2-3.tck"}
    ]
)
nv

NiiVue(height=300)

In [29]:
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}, 
        {"path": "../data/diffusion/cortex_together.mif"},
        {"path": "../data/diffusion/corpus_callosum.mif"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/fulltract_CC2-3.tck"}
    ]
)
nv.volumes[1].colormap='green'
nv.volumes[2].colormap='green'
nv

NiiVue(height=300)

### Once more, you can clean this up by drawing your own ROIs. If you want you can draw exclusion ROIs in <a href=https://mrtrix.readthedocs.io/en/latest/reference/commands/mrview.html>mrview</a> and then use tckedit with the -exclude option to exclude streamlines that cross your ROIs.

# Now let's try to divide this tract into three parts: a motor, somatosensory, and premotor part

**Here we use the same cortical ROIs as we used previously in the CST example.**

In [32]:
#!tck2connectome ../data/diffusion/fulltract_CC2-3.tck ../data/diffusion/three_regions_together_CC.mif ../data/diffusion/connectivity_values_CC_subdivision.csv -out_assignments ../data/diffusion/assignments_CC_subdivision.txt -symmetric -zero_diagonal -force
#!connectome2tck ../data/diffusion/fulltract_CC2-3.tck ../data/diffusion/assignments_CC_subdivision.txt ../data/diffusion/CC_subdivision -nodes 1,3,5 -force

tck2connectome: [WARNING] existing output files will be overwritten
tck2connectome: [100%] Constructing connectome[0K
connectome2tck: [WARNING] existing output files will be overwritten
connectome2tck: [done] reading streamline assignments file[0K
connectome2tck: [100%] Extracting tracks from connectome[0K


*Note: The image called three_regions_together_CC.mif has six values (1 to 6). Area 1, 3, 5 represent the left cortical ROIs (somatosensory, motor, and premotor, respectively), and area 2, 4, 6 represent right cortical ROIs (somatosensory, motor, and premotor, resp.). connectome2tck will export files that end on both sides for each pair of cortical ROIs, and add the suffix '1-2', '3-4', '5-5' to the exported filename (for somatosensory, motor, and premotor bundles, resp.).*

In [33]:
#Display the somatosensory part of the CC
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CC_subdivision1-2.tck"}
    ]
)
nv

NiiVue(height=300)

In [34]:
#Display the motor part of the CC
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CC_subdivision3-4.tck"}
    ]
)
nv

NiiVue(height=300)

In [35]:
#Display the premotor part of the CC
nv=NiiVue()
nv.load_volumes(
    [
        {"path": "../data/diffusion/anatomical.nii.gz"}
    ]
)
nv.load_meshes(
    [
        {"path": "../data/diffusion/CC_subdivision5-6.tck"}
    ]
)
nv

NiiVue(height=300)